### ISSUE TITLE
`verb` and `verb-cli`

### ISSUE THREAD

**Issue by nknapp** [verbose/verb#63]
I was wondering about the main reasons for separating `verb` and `verb-cli` (I could scan all closed issues, but there are quite a lot). I think the fact that `verb-cli@0.6.2` only works with `verb@0.7.x`, and `verb-cli@0.7.0` is needed for `verb@0.8.x` is not a good thing:
Hypothetically, I might have multiple repositories using different versions of `verb`. Some of them might be forks, so I cannot just change the `verb`-version in the `package.json` file. If I want to work with all these repos, I would need a version of `verb-cli` that works with both `verb@0.7.x` and `verb@0.8.x`. 

Thus, the second question, actually the more important one: What is responsible for these version-dependencies? 
From my point of view, it would be a good thing to put as much logic into `verb` and make `verb-cli` a wrapper so simple, that no future incompatibilities are possible:

* If a project has a `verbfile.js` (and thus a dependencies on `verb`). Then the call should be 
  delegated to the project's `verb`-version.
* If a project has no `verbfile.js`, the call should be delegated to `verb-cli`'s global `verb`-dependency.

I'm not sure, but I can imagine, that even the command-line-parsing could be done by `verb`. The advantage would be (as I said) that every version of `verb-cli` would work with every version of `verb`, so that I don't have to globally install a new version, just because a repo uses a newer version of `verb`.

But this would come pretty close to joining `verb` and `verb-cli` into  I guess there is at least good reason, why you separated them in the first place. Otherwise you wouldn't have done it. Thus my first question.

I would also like to know, why I have to do `npm cache clean && npm i -g verb-cli` in order to install a new version of `verb-cli`. Shouldn't npm just install the newer version, even without a `cache clean`?

 PS: I appreciate your quick answers. This one is *not* urgent, take your time. I'm merely curious.

**Comment by tunnckoCore** [comment]
The idea is separation of concerns in first place. In second, verb-cli does nothing more than parsing arguments and running `verb` internally, no matter it use only `.verb.md` or `verbfile.js`.

I think it's not needed to do `npm cache clean && npm i -g verb-cli` every time.

I dont know how @jonschlinkert use verb, but I cant even try it and start using it from few month, because the bugs.

**Comment by jonschlinkert** [comment]
> why I have to do npm cache clean && npm i -g verb-cli in order to install a new version of verb-cli

You don't necessarily "need" to do this, but it's recommended for any module if you have problems when you do an upgrade. Think of it like rebooting your computer, e.g. _before you assume it's a bug, reboot. if that doesn't work, it might be a bug_.

> I was wondering about the main reasons for separating verb and verb-cli

The main reason that verb-cli is needed is so that you don't need to install verb itself locally unless you have a `verbfile.js`. I've considered re-combining the projects, but I still think this is the way to go. The goal is to achieve what you mention later: "a wrapper so simple, that ~~no~~ future incompatibilities are [unlikely]". Keep in mind that verb is not at 1.0 yet, so unfortunately there will be API changes. When the API stabilizes, we won't have these issues anymore.

> but I cant even try it and start using it from few month, because the bugs.

Verb's number one deficiency is documentation for itself lol. Verb only generates documentation, unfortunately it doesn't write it for you too :)  In other words, I don't have the issues you're having @tunnckoCore, but it's not your fault, it's mine, and it's not because of bugs, it's because you're trying to do things with it that I don't do. 

Bottom line is that I know verb has some issues, but IMHO it's the most useful project I have, and no other project comes close to filling the same needs - so I'm motivated to fix the issues, it would be great if other saw the same value as me and wanted to help me get the API stabilized, docs written and bugs fixed.

**Comment by jonschlinkert** [comment]
thinking out load... the single biggest root cause of many of the issues we're having with verb and assemble is how options are handled in Template. 

we need to have the ability to set default options that are easily overridden - either by users or inheriting applications. This means that many of the options will need to be lazy-loaded or some other sane way of allowing options to be overridden.

**Comment by nknapp** [comment]
> but it's not your fault, it's mine, and it's not because of bugs, it's because you're trying to do things with it that I don't do.

Sounds like you think you are the only one who is actively using verb at the moment? Looking at the daily download stats on npm makes me think differently. I have done some research and found `verb` to be to be the best choice for generating documentation, as long as you want to generate `.md`-files. It can probably do much more than I need to do.

But I agree that documentation is missing. I would especially be interested in implementation documentation. At the moment, it's very hard for me to understand errors. I tried to debug #62 myself, but without some knowledge of the inside of work, this is really hard and time-consuming.

Btw. I accidentially created a [wiki-page](https://github.com/verbose/verb/wiki). I didn't know I could do that. I can't delete it...
Maybe the github-wiki is not a bad-idea to collect information that should be documented, with pointers to the actual sources.

**Comment by jonschlinkert** [comment]
> Sounds like you think you are the only one who is actively using verb at the moment?

well lol, honestly I was starting to get that impression. I hadn't looked at npm, but I will pay more attention to those stats. thanks.

> I accidentially created a wiki-page. I didn't know I could do that. I can't delete it

no worries! the settings allow anyone from the community to do that for a reason! :)

> At the moment, it's very hard for me to understand errors

agreed, errors need to be improved, and honestly I'm struggling with the best way to do that. what I'm thinking is that there needs to be an error module that has error methods for:

- plugins
- middleware
- helpers

thoughts?

**Comment by tunnckoCore** [comment]
> what I'm thinking is that there needs to be an error module that has error methods for

Yea.

> It can probably do much more than I need to do.

Yea. And the installation time... and the counting "small" dependencies. Yea really, they are small. Small, small, but they are 40+, lol, think it.

Because that, I'm thinking we should extract the thing that do the simple readme generation/documentation from `.verb.md` (+ few default helpers) to other package which will be simple CLI runner (`verb-readme`?). Or why not as gulp plugin with builtin helpers? `gulp-verb[-readme]`?

**Comment by tunnckoCore** [comment]
@jonschlinkert I think you lost the core idea, cuz that passion, lol :+1: Is it was "small and simple"? It isn't so small. Even template isn't small for me. But yea, they are absolutely cool.

**Comment by jonschlinkert** [comment]
> Because that, I'm thinking we should extract the thing that do the simple readme generation/documentation from .verb.md (+ few default helpers) to other package which will be simple CLI runner (verb-readme?

Lol, that's awesome. [like this](https://github.com/jonschlinkert/grunt-readme)? I remember when I thought it would be that easy. I'd be happy to continue the discussion on a different issue.

**Comment by jonschlinkert** [comment]
> And the installation time... and the counting "small" dependencies

also, fwiw, I think this will get better over time in general. nothing to do with my libs in particular, but network speed, and npm's speed are improving. Having small modules and a lot of small dependencies is a necessary side effect of the module pattern - and the penalty is that it takes a little longer to install modules that honor this pattern. verb-cli takes about 5 seconds to install for me. after that, it runs almost instantly. It's a tradeoff I'm more than happy to make.

**Comment by tunnckoCore** [comment]
:+1: 

> Lol, that's awesome. like this?

eeerm, grunt.. haha. I dont understand grunt-ish language, lol. And even, I never used it.

> I remember when I thought it would be that easy.

haha, Im sure it is easy.
`minimist` + `fs` + helpers for related, apidocs, copyright + data from pkg.json and `data-store` in which will populate global git configs from your git tools -- for `.verb.md` usage. And for `docs/`-style we'll need also `glob-fs` (for experiment) and `include` helper.

> I'd be happy to continue the discussion on a different issue.

where?

**Comment by jonschlinkert** [comment]
> i think here's the place.

maybe it is. 

> minimist + fs + helpers for related, apidocs, copyright + data from pkg.json and data-store in which will populate global git configs from your git tools -- for .verb.md usage. And for docs/-style we'll need also glob-fs (for experiment) and include helper.

Try this: https://github.com/jonschlinkert/readme-generator

**Comment by tunnckoCore** [comment]
@jonschlinkert wohoho, too fast!

Exactly, cool. Wasn't so "easy", huh? lol.

We can now continue there, yeah. I have some more ideas.

**Comment by jonschlinkert** [comment]
lol, no you're missing the point. The hard part is getting it to work for more than readmes, and making it work for all users, which means normalizing data from the user's environment and package.json so that templates are populated consistently and correctly. That template might fail 75% of the time, who knows?

It took me 15 min. to create that project, and it will generate a readme very quickly. But so what? Should I use that sometimes, just because it shaves 4 miliseconds from load time, and use verb when readme-generator isn't capable of doing what I need? why bother? I'll still use verb because I don't want a readme generator. I want a documentation generator that is capable of doing everything I need: 

- table of contents
- multi-file table of contents (index)
- API documentation with relative paths

etc. and sure, we can add these things to readme-generator, but then ... it would be verb :)

**Comment by jonschlinkert** [comment]
I'll just keep trying to improve verb, make it faster, etc. I've said it before, without any question verb is my favorite project because of how much faster it makes me at maintaining projects. I do a lot with verb that you probably aren't thinking about. this past weekend I published a bunch of ansi color repos (because I don't want to use chalk with it's 15 dependencies just to style one log statement). anyway, all of those repos were generated with verb....

**Comment by tunnckoCore** [comment]
> I want a documentation generator that is capable of doing everything I need:

All that you need is just helpers, actually. Helpers that you already extract to other awesome small libs like `utils`, `template-helpers` and the org `github.com/helpers`. Helpers that run through just template(s) - 95% of the time readme generation + index.js api and stuff like license/copyright which is handled by helpers again.

> The hard part is getting it to work for more than readmes, and making it work for all users, which means normalizing data from the user's environment and package.json so that templates are populated consistently and correctly.

mm.. cant see the hard thing. We can make lib that collect data from all of your git tools (data from global gitconfig and local git info), use `data-store` to save to global home, then merge using `extend-shallow` plus `load-pkg`.
But okey, we have different views and we will continue to different ways - it is normal. When I'm more free I'll do what I think the things should be in the `tunnckoCore/docks` package.

> this past weekend I published a bunch of ansi color repos

yea, i noticed that, lol

**Comment by nknapp** [comment]
When I read [this article about good documentation](https://jacobian.org/writing/great-documentation/), I was a bit discouraged about doing such things myself. It's a bit oppionated, I guess, but it still shows that writing great documentation is a really hard and time-consuming thing to do.

**Comment by jonschlinkert** [comment]
> it still shows that writing great documentation is a really hard and time-consuming thing to do.

indeed. this comment says it all "over the years I’ve spent helping to write Django’s docs." By comparison, I'm just getting started with verb. We'll have great documentation soon, I'm working on it ;)

> All that you need is just helpers, actually

Awesome, I can't wait to see how you allow custom delimiters and different template engines using helpers (this is an absolute requirement, given that the target audience is developers and many readmes have code examples that contain conflicting syntax, or templates that should not be rendered).  I look forward to seeing everything up and running ;)

**Comment by tunnckoCore** [comment]
> how you allow custom delimiters and different template engines using helpers

not with helpers, of course, but is pretty little job with `jstransformer` (for rendering) and `delims`, i guess.

> I look forward to seeing everything up and running

yea, I'll ping you one day, 100% :)

**Comment by jonschlinkert** [comment]
k, I'm going to close this. I'll continue to work on improving verb and verb-cli and the separation of concerns between the two. in the meantime, feel free to create a new issue if you have any other problems. I appreciate the feedback
